octree_web_viewer/src/backend.rs

```diff
@@ -65,82 +73,92 @@ fn pad(input: &mut Vec<u8>) {
-            reply_blob
-                .write_f32::<LittleEndian>(node_data.meta.bounding_cube.edge_length())
-                .unwrap();
-
-            // Number of points.
-            reply_blob
-                .write_u32::<LittleEndian>(node_data.meta.num_points as u32)
-                .unwrap();
-
-            // Position encoding.
-            let bytes_per_coordinate = node_data.meta.position_encoding.bytes_per_coordinate();
-            reply_blob.write_u8(bytes_per_coordinate as u8).unwrap();
-            assert!(
-                bytes_per_coordinate * node_data.meta.num_points as usize * 3
-                    == node_data.position.len()
-            );
-            assert!(node_data.meta.num_points as usize * 3 == node_data.color.len());
-            pad(&mut reply_blob);
-
-            reply_blob.append(&mut node_data.position);
-            pad(&mut reply_blob);
-
-            reply_blob.append(&mut node_data.color);
-            pad(&mut reply_blob);
-
-            num_nodes_fetched += 1;
-            num_points += node_data.meta.num_points;
-        }
-
-        let duration_ms = (time::precise_time_ns() - start) as f32 / 1000000.;
-        println!(
-            "Got {} nodes with {} points ({}ms).",
-            num_nodes_fetched, num_points, duration_ms
-        );
-
-        let content_type = "application/octet-stream".parse::<Mime>().unwrap();
-        Ok(Response::with((content_type, iron::status::Ok, reply_blob)))
+impl<S: 'static> Handler<S> for NodesData {
+    type Result = FutureResponse<HttpResponse>; //alias forBox<Future<Item=HttpResponse, Error=Error>>;
+
+    fn handle(&self, req: &HttpRequest<S>) -> Self::Result {
+        let octree = Arc::clone(&self.octree); //has to be moved into future
+        let message_body_future = Json::<Vec<String>>::extract(req).from_err();
+        message_body_future
+            .and_then(move |extract_result| {
+                let data: Vec<String> = Json::into_inner(extract_result);
+                let nodes_to_load = data
+                    .into_iter()
+                    .map(|e| octree::NodeId::from_str(e.as_str()));
+
+                // So this is godawful: We need to get data to the GPU without JavaScript herp-derping with
+                // it - because that will stall interaction. The straight forward approach would be to ship
+                // json with base64 encoded values - unfortunately base64 decoding in JavaScript yields a
+                // string which cannot be used as a buffer. So we would need to manually convert this into
+                // an Array with is very slow.
+                // The alternative is to binary encode the whole request and parse it on the client side,
+                // which requires careful constructing on the server and parsing on the client.
+                let start = time::precise_time_ns(); //timing is not comparable to the previous implementation in iron
```

Comment:
How about creating `start` in the first line of `handle` and create a future, which internally assigns `precise_time_ns()` to `start` and starts the future chain? Then timing would still be asynchronous, but at least covers everything including body parsing.

Reply:
done in pairing